### PR TITLE
Expand Held Items

### LINKED ISSUE #1076: [SUGGESTION] More Rare Hold Items/Drops
Right now, only 3 pokemon in the game have a rare Hold Item, effectively their drop when defeated. This makes the "+drop%" item almost useless except for the once or twice in the game when you'd want those drops, and you usually just buy the drops with Quest Points anyway. What if more pokemon had rare drops that inter-played with some other game features? (As they kinda do with Farm, and could do with Mining)

**Current Behavior**
Only 3 pokemon have rare hold items, of which are just evolution stones where it be easier to breed for a shiny then get enough stones to trade.

**Improved Behavior**
Introduce an entire list of potential rare hold items. Still have them be rare (5% is the canon rate for rare items and 50% for common ones, but this could be reduced) but at least make farming specific routes more of a tactical choice depending on what you want out of it. (Dungeon coins vs money vs achievements vs mining progress) This could even play in to some Quests like "Find X of Y item."

**Detailed Description**
The following is a list of every item already in PokeClicker, plus Nuggets, which could potentially be added to the Mining mini-game and "sold" for Coins instead of Diamonds. (Overall, having access to these drops would probably make the Mine easier, so this is technically a slight nerf to balance that slightly back towards center.) For a graphic, Iron Spheres are literally just recolored Nuggets, so coloring them gold again like how some plates were made by recoloring, would work just fine.

Moon Stone - Cleffa/Clefairy/Clefable/Lunatone
Sun Stone - Solrock
Nugget - Meowth/Grimmer/Muk/Trubbish/Garbodor (Minable item sold for pokecash?)
Kingstone - Poliwhirl/Poliwrath/Politoed/Hariyama/Hawlucha/Slowbro (already on the last one)
Everstone - Geodude/Graveler/Golem/Roggenrola/Boldore/Gigalith
Hardstone - Onyx/Corsola/Aron/Lairon/Aggron
Rarebone - Cubone/Marrowak/Dwebble/Crustle
Ovalstone - Chansey/Blissey/Happiny
Electrizer - Electrabuzz family line
Magmarizer - Magmar family line
Revive - Zigzagoon
Max Revive - Linoone
Lucky Egg - Pelipper
Lucky Magnet - Magneton/Probopass
Star Piece - Jirachi/Meloetta/Minior
Heart Scale - Luvdisk (OP? Maybe a unique 1% or 0.1% base drop rate?)

### PR DESCRIPTION
**Pull request by HLXII**
Closes #1076 

There were only three Held Items implemented, which made the mechanic not very useful.

Looking at the item drop table in canon, there are a lot more drops available. I've added some of these drops onto pokemon (if the item currently exists in Pokeclicker).

This change also requires the BagHandler API I originally created in the PokeLab branch, so I moved those changes onto this branch.

Here's the list of new item drops:

Held Item | Pokemon
------------ | -------------
Moon Stone | Cleffa, Clefairy, Clefable, Lunatone
Sun Stone | Solrock
King's Rock | Poliwhirl, Poliwrath, Politoed, Slowpoke, Slowbro, Sloking, Hariyama, Hawlucha
Metal Coat | Magnemite, Magneton, Magnezone, Steelix, Beldum, Metang, Metagross, Bronzor, Bronzong
Dragon Scale | Horsea, Seadra, Kingdra, Dratini, Dragonair, Dragonite, Bagon, Shelgon, Salamance
Electirizer | Elekid, Electabuzz, Electrivire
Magmarizer | Magby, Magmar, Magmortar
Deep Sea Scale | Chinchou, Lanturn, Gorebyss, Relicanth, Basculin (Blue)
Deep Sea Tooth | Carvanha, Sharpedo, Huntail, Basculin (Red)
Razor Fang | Bruxish
Razor Claw | Jangmo-o, Hakamo-o, Kommo-o
Upgrade | Porygon2
Iron Ball | Mawile
Heart Scale | Luvdisc
Rare Bone | Dwebble, Crustle, Cubone, Marowak
Everstone | Geodude, Roggenrolla, Boldore, Gigalith
Hard Stone | Graveler, Golem, Onix, Corsola, Nosepass, Probopass, Aron, Lairon, Aggron
Oval Stone | Chansey, Happiny, Blissey
Revive | Zigzagoon
Max Revive | Linoone
Star Piece | Jirachi, Meloetta, Minior
Lum Berry | Mew, Celebi, Shaymin

### CODE REVIEW

**Comment by TrueWolves** [comment]
If this goes through, you may want to close (and/or implement parts of) this suggestion that's been sitting for over a month now as well:
https://github.com/pokeclicker/pokeclicker/issues/1076
Looks like you hit most everything suggested there anyway.

**Comment by RedSparr0w** [comment]
I'm not sure about having the Underground items (diamond exchange items) dropping from such common Pokémon, Maybe just remove those from any of the Pokémon that can be found on routes, and it should be all good.

**Comment by HLXII** [comment]
> I'm not sure about having the Underground items _(specifically diamond exchange items)_ dropping from such common Pokémon, Maybe just remove those from any of the Pokémon that can be found on routes, and it should be all good.

Removing only from route Pokemon is a fragile solution, as it seems like we're adding old region pokemon into the new routes in a seemingly random pattern. We'd have to be careful in making sure the Underground item dropping pokemon never appear in routes, which would be annoying.

I'm toying with the idea of increasing `GameConstants.ROUTE_HELD_ITEM_CHANCE` to make item drops even rarer. That way player's would only get 1~2 of the item drops per day. Currently the calculation goes:

`1 / (ROUTE_HELD_ITEM_CHANCE / (1.5 if Item Magnet)) * (1 / Total Route Pokemon)`

So for Dratini in Route 45, we'd have:

`(1 / (512 / 1.5)) * (1 / 9) = 0.00032552083`

If the player has the Item Magnet on constantly, and they sit in the route, they'll defeat a Pokemon every second (assuming no active play). Using the binomial distribution, we can check for 84600 events with chance ~0.0033. With that, there's already like a 99% chance the player will get ~20 items, unless I'm missing something.

I think lowering the overall chances for item drops might make more sense, or even splitting the ROUTE_HELD_ITEM_CHANCE to have a lower chance for Underground items specifically?

It's a bit unfortunate that any new mechanics interacting with the Underground are blocked by its Diamond balancing.

**Comment by RedSparr0w** [comment]
>  it seems like we're adding old region pokemon into the new routes in a seemingly random pattern.

All the route encounters are just based on the original games data AFAIK

> I think lowering the overall chances for item drops might make more sense, or even splitting the ROUTE_HELD_ITEM_CHANCE to have a lower chance for Underground items specifically?

Yeah, I think that may be a better solution, the other items aren't as important IMO, but if you think they all need to be decreased I'm fine with that too.

**Comment by TrueWolves** [comment]
Honestly, 20 items per *24 hours* doesn't seem all that much, considering just how many more items can be mined with active effort.

File: src/scripts/underground/Underground.ts [hunk 1]
```diff
@@ -151,6 +151,14 @@ class Underground implements Feature {
         }
     }
 
+    public static getMineItemByName(name: string): UndergroundItem {
+        for (const item of UndergroundItem.list) {
+            if (item.name == name) {
+                return item;
+            }
+        }
```
Comment (RedSparr0w) [inline src/scripts/underground/Underground.ts:159]: Could we simplify this with a find?
```suggestion
        return UndergroundItem.list.find(i => i.name == name);
```

File: src/scripts/shards/Shards.ts [hunk 2]
```diff
@@ -124,6 +124,10 @@ class Shards implements Feature {
         }
     }
 
+    public static image(type: number): string {
+        return `assets/images/shards/${type}.png`;
+    }
+
```
Comment (RedSparr0w) [inline src/scripts/shards/Shards.ts:130]: Will this need to be updated with #1197 being merged?

Comment (HLXII) [reply src/scripts/shards/Shards.ts]: Yep, I've merged the changes

**Review by RedSparr0w — APPROVED** [review]
👍 Changes look good, thanks for the PR!

**Pull request merged**